### PR TITLE
Fixed #33937 -- Optimized serialization of related m2m fields without natural keys.

### PR DESCRIPTION
**Pull request by marknotfound**
Optimized m2m serialization when using a primary key by only retrieving the primary key from the related model rather than all fields.

### CODE REVIEW

File: django/core/serializers/python.py [hunk 1]
```diff
@@ -70,15 +70,22 @@ def handle_m2m_field(self, obj, field):
...
 
+                def queryset_iterator(obj, field):
+                    return getattr(obj, field.name).only("pk").iterator()
+
             m2m_iter = getattr(obj, "_prefetched_objects_cache", {}).get(
                 field.name,
-                getattr(obj, field.name).iterator(),
+                queryset_iterator(obj, field),
             )
+
```
Comment (felixxm) [inline django/core/serializers/python.py:88]: Please revert unrelated blank line.
```suggestion
```

File: django/core/serializers/python.py [hunk 2]
```diff
@@ -70,15 +70,22 @@ def handle_m2m_field(self, obj, field):
...
                 def m2m_value(value):
                     return self._value_from_field(value, value._meta.pk)
 
+                def queryset_iterator(obj, field):
+                    return getattr(obj, field.name).only("pk").iterator()
+
             m2m_iter = getattr(obj, "_prefetched_objects_cache", {}).get(
                 field.name,
-                getattr(obj, field.name).iterator(),
+                queryset_iterator(obj, field),
```
Comment (felixxm) [inline django/core/serializers/python.py:86]: We should add the same to the `django.core.serializers.xml_serializer.Serializer.handle_m2m_field`.

File: tests/serializers/tests.py [hunk 3]
```diff
@@ -410,6 +410,26 @@ def test_serialize_inherited_fields(self):
...
+
+    def test_serialize_no_only_with_natural_keys(self):
+        with mock.patch("serializers.models.Article.categories") as mock_categories:
+            serializers.serialize(
+                self.serializer_name,
+                Article.objects.all(),
+                use_natural_foreign_keys=True,
+            )
+            for categories in mock_categories:
+                categories.only.assert_not_called()
```
Comment (felixxm) [inline tests/serializers/tests.py:431]: I'd prefer not to use `mock` and check generated SQL instead. Also, `Category` doesn't have `natural_key()` so it works with `XML` serializer even when optimization is not use. Maybe:
```python
    def test_serialize_only_pk(self):
        with self.assertNumQueries(5) as ctx:
            serializers.serialize(
                self.serializer_name,
                Article.objects.all(),
                use_natural_foreign_keys=False,
            )
        categories_sql = ctx[1]["sql"]
        self.assertNotIn(connection.ops.quote_name("meta_data_id"), categories_sql)
        meta_data_sql = ctx[2]["sql"]
        self.assertNotIn(connection.ops.quote_name("kind"), meta_data_sql)

    def test_serialize_no_only_pk_with_natural_keys(self):
        with self.assertNumQueries(5) as ctx:
            serializers.serialize(
                self.serializer_name,
                Article.objects.all(),
                use_natural_foreign_keys=True,
            )
        categories_sql = ctx[1]["sql"]
        self.assertNotIn(connection.ops.quote_name("meta_data_id"), categories_sql)
        # CategoryMetaData with natural_key().
        meta_data_sql = ctx[2]["sql"]
        self.assertIn(connection.ops.quote_name("kind"), meta_data_sql)
```

**Review by felixxm — COMMENTED** [review]
@marknotfound Thanks for this patch :+1:

**Comment by marknotfound** [comment]
> @marknotfound Thanks for this patch 👍

@felixxm Thank you for your review! I've implemented your recommendations. I appreciate the guidance on testing the generated SQL. I was struggling a bit on the best way to test this change.

**Comment by felixxm** [comment]
@marknotfound Thanks :+1: Welcome aboard :boat:
